### PR TITLE
Rework ERT Tool

### LINKED ISSUE #910: Can't join ERT
Captain called an ERT, it self aborted due to not enough candidates

Called another one, people are unable to join, getting the error "You cannot join the response team at this time."

when using the IC > Join Response Team verb

### PR DESCRIPTION
**Pull request by DeadLineSS13**
## About this PR
Reworked ERT Tool.

## Changelog
🆑DeadLine
Fix: Fixed a ERT Tool. Now this good worked..
/🆑

### CODE REVIEW

File: code/datums/emergency_calls/kelion.dm
Comment (Terror4000rus) [inline code/datums/emergency_calls/kelion.dm]: It really doesn't sound like a good idea to give someone such famous names

File: html/changelogs/DeadLine.yml
Comment (Terror4000rus) [inline html/changelogs/DeadLine.yml]: But that is not your byond Ckey .-.
Your choice, btw, I guess

**Review by Terror4000rus — COMMENTED** [review]
Nice rework :D

File: code/__defines/flags.dm
Comment (DeadLineSS13) [inline code/__defines/flags.dm:61]: Flags for datums. (In future this need)

File: code/datums/emergency_calls/kelion.dm
Comment (JakeDaBoss) [inline code/datums/emergency_calls/kelion.dm]: Please refrain from adding lore names as characters.

File: code/datums/emergency_calls/kelion.dm
Comment (JakeDaBoss) [inline code/datums/emergency_calls/kelion.dm]: Idem.

File: code/_helpers/unsorted.dm
Comment (DeadLineSS13) [inline code/_helpers/unsorted.dm:1087]: Old version ref.

File: code/datums/emergency_calls/kelion.dm
Comment (JakeDaBoss) [inline code/datums/emergency_calls/kelion.dm]: Change this into CEC Headquarters.

File: code/datums/emergency_calls/kelion.dm
Comment (JakeDaBoss) [inline code/datums/emergency_calls/kelion.dm]: "You are part of the Kellion Repair Team"

File: code/datums/emergency_calls/emergency_call.dm
Comment (DeadLineSS13) [inline code/datums/emergency_calls/emergency_call.dm:6]: There you can create any ERT observing the template.

**Review by JakeDaBoss — CHANGES_REQUESTED** [review]
A few comments related to how things are displayed in the game.

File: html/changelogs/DeadLine.yml
Comment (DeadLineSS13) [reply html/changelogs/DeadLine.yml]: I like to call the author of changes with a GIT account more.

File: code/datums/emergency_calls/kelion.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/kelion.dm]: There are no variables that use this abbreviation. So it's good to call things own things.

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm:12]: A specific requirement here:

We need any messages about requesting/dispatching emergency teams, to be shown to all crewmembers, but NOT to be shown to necromorphs and signals

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: Did you add this variable? If so, it seems a bit ambiguously named

if not, then its probably intended for antag selection, this might not be a good way to use it

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: don't check mind twice, do this one first

Comment (NanakoAC) [reply code/datums/emergency_calls/emergency_call.dm]: also, use get_mind instead of .mind

File: code/datums/emergency_calls/kelion.dm
Comment (NanakoAC) [reply code/datums/emergency_calls/kelion.dm]: what famous names, you mean kellion? That was already planned long ago, these ERTs existed in code before this PR

File: code/datums/emergency_calls/unitologi.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/unitologi.dm]: forcibly giving people a random name is odd, do they have a way to change it?

File: code/game/antagonist/outsider/ert.dm
Comment (NanakoAC) [inline code/game/antagonist/outsider/ert.dm:118]: i hope all this stuff has been transfererred over

File: code/datums/emergency_calls/unitologi.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/unitologi.dm]: We can give them the opportunity to change the Name.

File: code/datums/emergency_calls/emergency_call.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/emergency_call.dm]: Add.

File: code/game/antagonist/outsider/ert.dm
Comment (DeadLineSS13) [reply code/game/antagonist/outsider/ert.dm:118]: Yep.

File: code/datums/emergency_calls/emergency_call.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/emergency_call.dm]: This check mind. Because M.mind && (M.mind in candidates) can TRUE with mind = null.

File: code/game/gamemodes/game_mode.dm
Comment (NanakoAC) [inline code/game/gamemodes/game_mode.dm]: I don't believe DM actually supports typed lists
this varname is too ambiguous.
And we usually managed globals like baycode

Maybe do GLOBAL_LIST_EMPTY(all_distress_call_datums)

Comment (DeadLineSS13) [reply code/game/gamemodes/game_mode.dm]: This variable is only used in ERT and DM can still use list datum, so it is better to ignore the use of global variables.

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm:4]: use managed globals for these. See the files in _global_vars for examples

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm]: picked call is too ambiguous

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm:55]: Removing this proc? Does that mean joining an ERT is automatic based on antag preferences now?

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm]: Maximum team size instead of number of mobs

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm:70]: ah i see you just moved this.

Honestly though, remove it. Antag settings are enough opting in, we dont need people to manually click a verb every round

File: html/changelogs/DeadLine.yml
Comment (NanakoAC) [reply html/changelogs/DeadLine.yml]: then use your full git account name,  DeadLineSS13

**Review by NanakoAC — CHANGES_REQUESTED** [review]
Lots of small concerns mostly about coding practises and necromorph message visibility

File: code/game/response_team.dm
Comment (DeadLineSS13) [reply code/game/response_team.dm:55]: I did not delete it, you can view file.

File: code/game/response_team.dm
Comment (DeadLineSS13) [reply code/game/response_team.dm]: Don't forget that Emergency call support all spawn mobs. So it's not just people who can be in ERT.

**Comment by DeadLineSS13** [comment]
image
Join Response Clicable.

File: code/_macros.dm
Comment (NanakoAC) [inline code/_macros.dm]: we have datum.is_necromorph(), use that

If you absolutely need a macro version, make it a wrapper to call that.
It's much more complex, and covers a variety of odd cases like divider heads taking over a body

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: call these members_max and members_min

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: byond doesnt support typed lists afaik?

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: call this weight, not probability

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm]: this name is grammatically strange, what is this list for?

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [reply code/datums/emergency_calls/emergency_call.dm]: check mind once, after client.

then the second check if mind is in candidates

File: code/datums/emergency_calls/unitologi.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/unitologi.dm]: five different unitologist outfits were already defined in another file

File: code/datums/emergency_calls/unitologi.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/unitologi.dm]: call this /datum/emergency_call/deliverance

Also the name of this file is wrong

File: code/game/response_team.dm
Comment (NanakoAC) [reply code/game/response_team.dm]: team size doesnt necessarily imply people

**Review by NanakoAC — CHANGES_REQUESTED** [review]
various minor concerns about variable, file and typenames

File: code/datums/emergency_calls/emergency_call.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/emergency_call.dm]: > this name is grammatically strange, what is this list for?

This name for specials members of team ERT. Example medic, engi and etc.

File: code/datums/emergency_calls/unitologi.dm
Comment (DeadLineSS13) [reply code/datums/emergency_calls/unitologi.dm]: > five different unitologist outfits were already defined in another file

Already used.

File: code/game/response_team.dm
Comment (DeadLineSS13) [reply code/game/response_team.dm]: > picked call is too ambiguous

We drop the old call just to be sure.

**Review by JakeDaBoss — APPROVED** [review]
Inherent spelling and all that seems fine now. @NanakoAC waiting for your sign-off on the code part.

File: code/_macros.dm
Comment (NanakoAC) [inline code/_macros.dm:68]: why remove this?

File: code/datums/emergency_calls/deliverance.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/deliverance.dm]: weight not weigh

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [reply code/datums/emergency_calls/emergency_call.dm]: why is this tied to gamemode at all? What if we want to do multiple emergency calls?

File: code/datums/emergency_calls/emergency_call.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/emergency_call.dm:113]: this whole proc seems overlarge and monolithic

File: code/datums/emergency_calls/deliverance.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/deliverance.dm:12]: why is all of this stuff in the deliverance subtype? Most of this seems like generic behaviour. Mind transfer, outfit picking, skills, namechanging, etc

File: code/datums/emergency_calls/kelion.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/kelion.dm:12]: this is a crazy degree of copypasting code

File: code/datums/emergency_calls/usm.dm
Comment (NanakoAC) [inline code/datums/emergency_calls/usm.dm:12]: And again, copypasting. This isn't how overriding should be done, this behaviour needs to be in base class and controlled by variables

File: code/datums/mind.dm
Comment (NanakoAC) [inline code/datums/mind.dm:76]: what is this new variable needed for?

File: code/datums/mind.dm
Comment (NanakoAC) [inline code/datums/mind.dm]: check not chec

File: code/game/gamemodes/game_mode.dm
Comment (NanakoAC) [inline code/game/gamemodes/game_mode.dm]: I think i mentioned before these varnames are too generic,

File: code/game/response_team.dm
Comment (NanakoAC) [reply code/game/response_team.dm]: its still called picked_call, its a generic varname which isn't sufficiently self documenting

File: code/game/response_team.dm
Comment (NanakoAC) [inline code/game/response_team.dm]: of not off

File: code/game/response_team.dm
Comment (NanakoAC) [reply code/game/response_team.dm:70]: This verb is still here, this isn't resolved

File: code/modules/client/preference_setup/antagonism/01_candidacy.dm
Comment (NanakoAC) [inline code/modules/client/preference_setup/antagonism/01_candidacy.dm:5]: this doesn't need to be an option, just selecting candidacy for the ERT roles is enough

File: code/procs/announce.dm
Comment (NanakoAC) [inline code/procs/announce.dm]: is_necromorph will return true for signals as well, so issignal is not needed here

**Review by NanakoAC — CHANGES_REQUESTED** [review]
Many small concerns, a few larger concerns with copypasting and lack of modularity, needs refactoring

File: code/datums/mind.dm
Comment (DeadLineSS13) [reply code/datums/mind.dm:76]: > what is this new variable needed for?

For line 101.

File: code/_macros.dm
Comment (DeadLineSS13) [reply code/_macros.dm:68]: > why remove this?

Check line 58

**Comment by JakeDaBoss** [comment]
@DeadLineSS13 Was this PR finished and for review?

**Comment by DeadLineSS13** [comment]
> @DeadLineSS13 Was this PR finished and for review?

Yep. But this thing need massive tests.

**Pull request closed without merging**
